core/sail/memory/src/main/java/org/eclipse/rdf4j/sail/memory/MemorySailStore.java

```diff
@@ -976,80 +965,41 @@ public CloseableIteration<? extends Resource, SailException> getContextIDs() thr
 			// Create a list of all resources that are used as contexts
 			ArrayList<MemResource> contextIDs = new ArrayList<>(32);
 
-			Lock stLock = openStatementsReadLock();
-			try {
-				int snapshot = getCurrentSnapshot();
-				try (WeakObjectRegistry.AutoCloseableIterator<MemIRI> memIRIsIterator = valueFactory
-						.getMemIRIsIterator()) {
-					while (memIRIsIterator.hasNext()) {
-						MemResource memResource = memIRIsIterator.next();
-						if (isContextResource(memResource, snapshot)) {
-							contextIDs.add(memResource);
-						}
+			int snapshot = getCurrentSnapshot();
+
+			try (WeakObjectRegistry.AutoCloseableIterator<MemIRI> memIRIsIterator = valueFactory
+					.getMemIRIsIterator()) {
+				while (memIRIsIterator.hasNext()) {
+					MemResource memResource = memIRIsIterator.next();
+					if (isContextResource(memResource, snapshot)) {
+						contextIDs.add(memResource);
 					}
 				}
+			}
 
-				try (WeakObjectRegistry.AutoCloseableIterator<MemBNode> memBNodesIterator = valueFactory
-						.getMemBNodesIterator()) {
-					while (memBNodesIterator.hasNext()) {
-						MemResource memResource = memBNodesIterator.next();
-						if (isContextResource(memResource, snapshot)) {
-							contextIDs.add(memResource);
-						}
+			try (WeakObjectRegistry.AutoCloseableIterator<MemBNode> memBNodesIterator = valueFactory
+					.getMemBNodesIterator()) {
+				while (memBNodesIterator.hasNext()) {
+					MemResource memResource = memBNodesIterator.next();
+					if (isContextResource(memResource, snapshot)) {
+						contextIDs.add(memResource);
 					}
 				}
-			} finally {
-				stLock.release();
 			}
 
 			return new CloseableIteratorIteration<>(contextIDs.iterator());
 		}
 
 		@Override
-		public CloseableIteration<? extends Statement, SailException> getStatements(Resource subj, IRI pred, Value obj,
+		public CloseableIteration<MemStatement, SailException> getStatements(Resource subj, IRI pred, Value obj,
 				Resource... contexts) throws SailException {
-			if (!explicit && !mayHaveInferred && snapshot >= 0) {
-				return EMPTY_ITERATION;
-			}
-
-			Lock stLock = openStatementsReadLock();
-			try {
-				CloseableIteration<MemStatement, SailException> statementIterator = createStatementIterator(subj, pred,
-						obj, explicit, getCurrentSnapshot(), contexts);
-				try {
-					return LockingIteration.getInstance(stLock, statementIterator);
-				} catch (Throwable t) {
-					statementIterator.close();
-					throw t;
-				}
-			} catch (Throwable t) {
-				stLock.release();
-				throw t;
-			}
-
+			return createStatementIterator(subj, pred, obj, explicit, getCurrentSnapshot(), contexts);
 		}
```

Comment:
All operations agains the MemStatementList can now be free og locking and synchronization. Above is a read operation.